### PR TITLE
Prune workflows based on changed files

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
## Description
Only run tests based on things that have actually changed. For example, if only Python files have changed, we don't need to run the C++ tests.

Contributes to https://github.com/rapidsai/build-planning/issues/94

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: .github/workflows/pr.yaml [hunk 1]
```diff
@@ -10,6 +10,24 @@ concurrency:
...
+      test_python: ${{ steps.changed-files.outputs.ci_any_changed == 'true' || steps.changed-files.outputs.cpp_any_changed == 'true' || steps.changed-files.outputs.python_any_changed == 'true' }}
+    steps:
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+        with:
+          files_yaml: |
+            ci:
+              - ci/**
+            cpp:
```
Comment (robertmaynard) [inline .github/workflows/pr.yaml:32]: changes to root level `rapids_config.cmake`, and `build.sh` need to trigger tests

**Comment by vyasr** [comment]
@KyleFromNVIDIA do you want to make this a draft until you're done editing? I assume you have some more iterating to do.

File: .github/workflows/pr.yaml [hunk 2]
```diff
@@ -10,6 +10,55 @@ concurrency:
...
+        with:
+          ref: ${{ inputs.sha }}
+          persist-credentials: false
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+        with:
+          files_yaml: |
+            ci:
+              - ci/**
```
Comment (bdice) [inline .github/workflows/pr.yaml:40]:
```suggestion
              - .github/**
              - ci/**
```

**Review by bdice — COMMENTED** [review]
Almost any of the other directories (`.devcontainer`, `.github`, `ci`, `conda`, `notebooks`) or top-level files (`codecov.yml`, `dependencies.yaml`, `pyproject.toml`) can affect the behavior of C++ and/or Python tests. We also have a lot of possible cases where new files could affect the behavior of C++/Python tests.

I would propose a different strategy that is based on skipping tests when we know things are safe rather than allowing tests to run on only certain changes.

I'd propose this:
- Run Python tests always _unless_ changes were confined to the following paths (conjuction is OR):
  - `docs/`
  - `java/`
- Run C++ tests always _unless_ changes were confined to the following paths (conjunction is OR):
  - `docs/`
  - `java/`
  - `notebooks/`
  - `python/`

File: .github/workflows/pr.yaml (hunk 2 above)
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: I've added negative selectors based on the suggestions you made above instead.

**Comment by KyleFromNVIDIA** [comment]
We are now at the point where this new workflow can detect if there is any *net* change in the PR that should result in the corresponding tests being run:

https://github.com/rapidsai/cudf/actions/runs/10513648776/job/29129711025?pr=16642

Now, even if the PR contains commits that change the relevant directories, if later commits change them back, then the relevant tests no longer run:

https://github.com/rapidsai/cudf/actions/runs/10513660677/job/29129752104?pr=16642

Likewise, if a commit doesn't change the relevant directories, but an earlier commit did, then the tests are still run:

https://github.com/rapidsai/cudf/actions/runs/10513720481/job/29129943949?pr=16642

File: .github/workflows/pr.yaml [hunk 3]
```diff
@@ -55,9 +88,10 @@ jobs:
       build_type: pull-request
       enable_check_symbols: true
   conda-cpp-tests:
-    needs: conda-cpp-build
+    needs: [conda-cpp-build, changed-files]
     secrets: inherit
     uses: rapidsai/shared-workflows/.github/workflows/conda-cpp-tests.yaml@branch-24.10
+    if: needs.changed-files.outputs.test_cpp == 'true'
```
[the remark this answers is not in the document]
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml:100]: Let me do a quick test and find out

[the remark this answers is not in the document]
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: If the job is skipped, `pr-builder` is never run (it gets skipped). How does that impact merging?

[the remark this answers is not in the document]
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: Going based on https://stackoverflow.com/questions/65844033/conditional-needs-in-github-actions:

We can force `pr-builder` to run by giving it `if: always()`, but this means it passes even if one of the requisite jobs fails. We need a way to make `pr-builder` fail if one of the requisite jobs fails.

[the remark this answers is not in the document]
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: I've modified `pr-builder` to check the `needs` input to see if any jobs failed. PR incoming.

[the remark this answers is not in the document]
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: https://github.com/rapidsai/shared-workflows/pull/237

File: .github/workflows/pr.yaml [hunk 4]
```diff
@@ -36,6 +37,61 @@ jobs:
...
+          base_sha: ${{ fromJSON(steps.get-pr-info.outputs.pr-info).base.sha }}
+          files_yaml: |
+            cpp:
+              - '**'
+              - '!docs/**'
+              - '!java/**'
+              - '!notebooks/**'
+              - '!python/**'
+              # TODO: Remove this before merging
+              - '!.github/**'
```
Comment (bdice) [inline .github/workflows/pr.yaml:74]: Reminder comment to remove these before merging.

**Review by bdice — APPROVED** [review]
A few notes:
- I think `pandas-test-diff` will be safely skipped if its dependency on `pandas-tests` is skipped. That's the only case where we have added an `if changed` condition to a job that has downstream dependencies, so just wanted to make sure that seems safe to you too.
- We could be more aggressive in the future with Python job dependencies. For example, changes in `cudf` or `dask-cudf` shouldn't trigger `cudf-polars` tests, and changes in `cudf-polars` shouldn't trigger `cudf` or `pandas` tests. However I think we are doing enough in this PR to make a dent in CI usage and I don't want to add significantly more complication at once.

**Comment by KyleFromNVIDIA** [comment]
https://github.com/rapidsai/cudf/actions/runs/10527602668?pr=16642 has passed with all of the tests jobs being skipped. `pr-builder` has run successfully. I'm now going to try deliberately inducing a failing test.

**Comment by KyleFromNVIDIA** [comment]
The notebook test didn't fail because I happened to pick the *one* notebook that wasn't actually being run.

However, devcontainers happened to fail due to the recent merge of https://github.com/rapidsai/devcontainers/pull/271 (and this PR not having https://github.com/rapidsai/cudf/pull/15483), so we were able to get a failed `pr-builder` anyway.

I'm happy with the state of this PR. I'll remove the temporary modifications and test it one more time.

**Review by vyasr — APPROVED** [review]
This looks great. I agree with using negative rather than positive filtering. We should be conservative to make sure we don't lose test coverage. We can always filter more aggressively over time.

File: .github/workflows/pr.yaml [hunk 5]
```diff
@@ -37,6 +38,52 @@ jobs:
...
+          ref: ${{ inputs.sha }}
+          fetch-depth: ${{ fromJSON(steps.get-pr-info.outputs.pr-info).commits }}
+          persist-credentials: false
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+        with:
+          base_sha: ${{ fromJSON(steps.get-pr-info.outputs.pr-info).base.sha }}
+          files_yaml: |
+            cpp:
```
Comment (robertmaynard) [inline .github/workflows/pr.yaml:68]: nit: I believe we can exclude the img folder as well from `cpp`, `python`, and `java`.

File: .github/workflows/pr.yaml [hunk 6]
```diff
@@ -37,6 +38,52 @@ jobs:
...
+        with:
+          ref: ${{ inputs.sha }}
+          fetch-depth: ${{ fromJSON(steps.get-pr-info.outputs.pr-info).commits }}
+          persist-credentials: false
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+        with:
+          base_sha: ${{ fromJSON(steps.get-pr-info.outputs.pr-info).base.sha }}
+          files_yaml: |
```
Comment (robertmaynard) [inline .github/workflows/pr.yaml:67]: In general can we exclude the `README.md` and `CONTRIBUTING.md` from build runs? That would help reduce the amount of CI runs for pure doc updates

**Review by robertmaynard — APPROVED** [review]

**Pull request merged**